src/wbase/SendChannelShared.h

```diff
@@ -0,0 +1,210 @@
+
+    static std::atomic<uint64_t> scsSeqId; ///< Source for unique _scsId numbers
+
+    /// To help ensure that _streamMutex is locked before calling,
+    /// many member functions require a StreamGuard argument.
+    using StreamGuard = std::lock_guard<std::mutex> const&;
+
+    SendChannelShared()=delete;
+    SendChannelShared(SendChannelShared const&) = delete;
+    SendChannelShared& operator=(SendChannelShared const&) = delete;
+
+    static Ptr create(SendChannel::Ptr const& sendChannel,
+                      std::shared_ptr<wcontrol::TransmitMgr> const& transmitMgr);
+
+    ~SendChannelShared();
+
+    /// Wrappers for SendChannel public functions that may need to be used
+    /// by threads.
+    /// @see SendChannel::send
+    bool send(StreamGuard sLock, char const* buf, int bufLen) {
+        return _sendChannel->send(buf, bufLen);
+    }
+
+    /// @see SendChannel::sendError
+    bool sendError(StreamGuard sLock, std::string const& msg, int code) {
+        return _sendChannel->sendError(msg, code);
+    }
+
+    /// @see SendChannel::sendFile
+    bool sendFile(StreamGuard sLock, int fd, SendChannel::Size fSize) {
+        return _sendChannel->sendFile(fd, fSize);
+    }
+
+    /// @see SendChannel::sendStream
+    bool sendStream(StreamGuard sLock, xrdsvc::StreamBuffer::Ptr const& sBuf, bool last, int scsSeq=-1) {
+        return _sendChannel->sendStream(sBuf, last, scsSeq);
+    }
+
+    /// @see SendChannel::kill
+    bool kill(StreamGuard sLock, std::string const& note);
+
+    /// @see SendChannel::isDead
+    bool isDead() {
+        if (_sendChannel == nullptr) return true;
+        return _sendChannel->isDead();
+    }
+
+
+    /// Set the number of Tasks that will be sent using this SendChannel.
+    /// This should not be changed once set.
+    void setTaskCount(int taskCount);
+
+
+    /// Try to transmit the data in tData.
+    /// If the queue already has at least 2 TransmitData objects, addTransmit
+    /// may wait before returning. Result rows are read from the
+    /// database until there are no more rows or the buffer is
+    /// sufficiently full. addTransmit waits until that buffer has been
+    /// sent to the czar before reading more rows. Without the wait,
+    /// the worker may read in too many result rows, run out of memory,
+    /// and crash.
+    ///  @see SendChannelShared::_transmit code for further explanation.
+    bool addTransmit(bool cancelled, bool erred, bool last, bool largeResult,
+                     TransmitData::Ptr const& tData, int qId, int jId);
+
+    ///
+    /// @return true if inLast is true and this is the last task to call this
+    ///              with inLast == true.
+    /// The calling Thread must hold 'streamMutex' before calling this.
+    bool transmitTaskLast(StreamGuard sLock, bool inLast);
+
+    /// streamMutex is used to protect _lastCount and messages that are sent
+    /// using SendChannelShared.
+    std::mutex streamMutex;
```

Comment:
I have my doubts as well. So I'll look closer. The original code required it to keep the buffers in order. Code has moved and changed since then, so I don't think it needs to be outside, but there may still be cases where data is added and the mutex would be needed to keep the data in order. 

@iagaponenko is fond of passing lock_guards into functions that require mutex locks as an indication that the lock is held. That has shortcomings, such as there's no proof that the lock_guard has the correct mutex. lock_guards and mutex are not genric, if  there's one case where you need a unique_lock, all the locks have to be unique_lock or you can't pass it as a parameter. If a normal mutex doesn't work, you can have an issue. If the parent class requires lock_guard, and child class needs a recursive mutex and an occasional unique_lock, it wont be pretty.
Basically, it's just an indication, not proof, and it locks you into one type of lock+mutex. If you need a try_lock or something odd, it doesn't cope well.

Reply:
I think the patter that Igor (and me) uses applies to private methods, in your own class you certainly have more control over what kind of lock to use. If you expose a mutex as public then this pattern is certainly questionable.